### PR TITLE
Variable elimination

### PR DESCRIPTION
**Pull request by francescoTheSantis**
In this PR the following things have been implemented:
- variable elimination algorithm
- negative log-likelihood loss
- factor (different from parametric factor) which is used to handle intermediate factors that are obtained during inference
- (minor) allow to set prior distribution when generating a synthetic dataset using categorical_toy_dag.py

### CODE REVIEW

File: torch_concepts/nn/modules/high/base/model.py [hunk 1]
```diff
@@ -449,15 +455,21 @@ def filter_output_for_loss(self, forward_out, target):
         dict
             Dict with 'input' and 'target' for loss computation.
         """
+        if isinstance(forward_out, dict):
+            return {'input': forward_out['log_joint'], 'target': target}
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/base/model.py]: we must find a way not to hard-code 'log_joint' here

File: torch_concepts/nn/modules/high/base/model.py [hunk 2]
```diff
@@ -466,6 +478,8 @@ def filter_output_for_metrics(self, forward_out, target):
         dict
             Dict with 'preds' and 'target' for metric computation.
         """
+        if isinstance(forward_out, dict):
+            return {'preds': forward_out['logits'], 'target': target}
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/base/model.py]: same here

File: torch_concepts/nn/modules/mid/models/variable.py [hunk 3]
```diff
@@ -14,6 +14,11 @@
 
 from .....distributions import Delta
 
+# Distribution type groups.
+_BINARY_DISTRIBUTIONS = {Bernoulli, RelaxedBernoulli}
+_CATEGORICAL_DISTRIBUTIONS = {Categorical, OneHotCategorical, RelaxedOneHotCategorical}
+_CONTINUOUS_DISTRIBUTIONS = {Normal, MultivariateNormal, Delta}
```
Comment (gdefe) [inline torch_concepts/nn/modules/mid/models/variable.py:20]: is this safe to have such a limited pool of distributions here? what would happen if a user asks for a different distribution? would the code raise error?

File: torch_concepts/nn/modules/loss.py [hunk 4]
```diff
@@ -625,4 +625,51 @@ def forward(
...
+    def __init__(self):
+        super().__init__()
+        self._bce_fallback = nn.BCEWithLogitsLoss()
+
+    def forward(self, **kwargs) -> torch.Tensor:
+        input = kwargs['input']
+        target = kwargs['target']
+
+        # Flat 2D tensor → fallback to BCE (eval inference path)
+        if input.ndim == 2:
```
Comment (gdefe) [inline torch_concepts/nn/modules/loss.py:668]: I am not sure of this. why would I need a different behavior for evaluation? a loss should not be aware of this

File: torch_concepts/nn/modules/loss.py [hunk 5]
```diff
@@ -625,4 +625,51 @@ def forward(
             mask = torch.isfinite(input)
         if mask.any():
             return self.scale * input[mask].abs().mean()
-        return torch.tensor(0.0, device=input.device)
\ No newline at end of file
+        return torch.tensor(0.0, device=input.device)
+
+
+class JointNLLLoss(nn.Module):
```
Comment (gdefe) [inline torch_concepts/nn/modules/loss.py:629]: can we use a torch class for NLL loss?

this sounds to me too hard coded for the inferences that return log_joint

File: torch_concepts/nn/modules/loss.py [hunk 6]
```diff
@@ -625,4 +625,51 @@ def forward(
...
+
+    Returns:
+        torch.Tensor: Scalar NLL loss.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._bce_fallback = nn.BCEWithLogitsLoss()
+
+    def forward(self, **kwargs) -> torch.Tensor:
```
Comment (gdefe) [inline torch_concepts/nn/modules/loss.py:663]: the loss signature should be explicit. 

Right now, the signature of the dict after filter_output_for_loss should match the forward signature

**Comment by francescoTheSantis** [comment]
Outdated
